### PR TITLE
feat: add function to fetch user playtime

### PR DESCRIPTION
**Pull request by TheYuriG**
This is my first time making a module PR and a typescript PR so you can bet that I had *a lot of fun* figuring out why stuff didn't work.

```
const {
	exchangeNpssoForCode,
	exchangeCodeForAccessToken,
	makeUniversalSearch,
	getUserPlayedTime,
} = require('psn-api');
const fs = require('fs');

async function thisWillTestPlaytime() {
	const access = await exchangeNpssoForCode(
		'PUT-YOUR-NPSSO-CODE-HERE'
	);
	const auth = await exchangeCodeForAccessToken(access);
	const userList = await makeUniversalSearch(auth, 'PUT-THE-PSNID-YOU-WANT-TO-TEST-HERE', 'SocialAllAccounts');
	const userId = userList.domainResponses[0].results[0].socialMetadata.accountId;
	const timePlayed = await getUserPlayedTime(auth, userId);
	fs.writeFileSync('./timePlayed.json', JSON.stringify(timePlayed));
}

thisWillTestPlaytime();
```

Expect best practices not being followed and edits needing to be made, I punched way above my weight with this PR.
Also, I didn't create a test, but maybe the repro code above will ease that process.

### CODE REVIEW

**Comment by wescopeland** [comment]
I have adjusted the title to enforce a minor release. Releases are automated and entirely derived from commit messages. Check https://github.com/semantic-release/semantic-release and https://commitizen-tools.github.io/commitizen/tutorials/writing_commits/

File: src/models/game-media.model.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export interface Media {
```
Comment (wescopeland) [inline src/models/game-media.model.ts:1]:
```suggestion
export interface GameMedia {
```

Export names should match file names. This provides a better clue as to what is in the file before even opening it.

File: src/models/game-media.model.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  audios: any[];
+  videos: any[];
```
Comment (wescopeland) [inline src/models/game-media.model.ts:3]: `any` is dangerous, especially as we aspire to be the de-facto reference implementation for all of our consumers 😅

Let's ensure these are strongly-typed.

File: src/models/game-media.model.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  images: [
+    {
+      url: string;
+      format: string;
+      type: string;
+    }
+  ];
```
Comment (wescopeland) [inline src/models/game-media.model.ts:10]: This means `images` is a singleton containing an object with these three properties. I have not yet run the code so I am assuming this is a mistake (feel free to correct me if it is not!). You may be looking for:

```ts
images: Array<{
  url: string;
  format: string;
  type: string;
}>
```

Even better if we can also strongly type `format` and `type` -- I wonder if there is a known quantity for these props that is more specific than `string`.

File: src/models/title-concept.model.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import type { Media } from "./game-media.model";
+
+export interface GameConcept {
```
Comment (wescopeland) [inline src/models/title-concept.model.ts:3]:
```suggestion
export interface TitleConcept {
```

To match the file name.

File: src/models/user-played-time-response.model.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import type { GameConcept } from "./title-concept.model";
+import type { Media } from "./game-media.model";
+
+export interface GetUserPlayedTimeResponse {
+  /** The title ID of this game 
+      @example CUSA26431_00
+      */
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:7]:
```suggestion
  /** 
   * The title ID of this game 
   * @example CUSA26431_00
   */
```

This format better follows JSDoc standards for multi-line documentation comments.

File: src/models/user-played-time-response.model.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  titleId: string;
+
+  /** Default name for the title */
+  name: string;
+
+  /** Localized name for the title */
+  localizedName: string;
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:14]: Examples for these two would also be very helpful. This is a bit ambiguous as-is.

File: src/models/user-played-time-response.model.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  /** URL to game's console launch screen, from image.api.playstation.com */
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:16]: "from X" can be cut. Much more valuable would be an @example, which would demonstrate this.

File: src/models/user-played-time-response.model.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  imageUrl: string;
+
+  /** URL to localized game's console launch screen. Will default to imageUrl if not unique */
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:19]: This comment is good but could also use an example.

File: src/models/user-played-time-response.model.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  localizedImageUrl: string;
+
+  /** Type of game being played. Can be ps5_native_game, ps4_game or none */
+  category: string;
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:23]:
```ts
category: "ps5_native_game" | "ps4_game" | "none";
```

Is this an improvement?

File: src/models/user-played-time-response.model.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  /** Service where this game was acquired from: EA Access, none(purchased/none_purchased, ps_plus, other, */
+  service: string;
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:26]: I am wondering if we can strongly type this as well based on the comment.

File: src/models/user-played-time-response.model.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+  /** Metadata about this game, including titleIds and localization for other regions, genres, name and id */
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:28]:
```suggestion
  /** Metadata about this game */
```

We can omit this description - the type is much more concrete!

File: src/models/user-played-time-response.model.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  concept: GameConcept;
+
+  /** An object with audio, video, images properties that will contain arrays of objects with URL, format and type properties */
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:31]: Similar to the above, this can be much more generic. If folks are really interested they can go directly to the concrete type.

File: src/models/user-played-time-response.model.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  media: Media;
+
+  /** A Date object in ISO format */
+  firstPlayedDateTime: Date;
+
+  /** A Date object in ISO format */
+  lastPlayedDateTime: Date;
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:38]: I am confused by these. Are these ISO strings or actual date objects? We can strongly type both but native `Date` objects are their own thing.

File: src/models/user-played-time-response.model.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+   * @example `"PT11H16M43S"`
+   */
+  playDuration: string;
```
Comment (wescopeland) [inline src/models/user-played-time-response.model.ts:43]: The example is good but we definitely need to elaborate on what this example means so our comment can serve as a reference.

File: src/user/USER_TIME_PLAYED_URL.ts [hunk 15]
```diff
@@ -0,0 +1,2 @@
+export const USER_TIME_PLAYED_URL =
```
Comment (wescopeland) [inline src/user/USER_TIME_PLAYED_URL.ts:1]: It seems to me this isn't a users-related URL. It's a gamelist-related URL. I think all of this new code should live under a new _src/gamelist_ directory. Then this variable can be renamed to `GAMELIST_BASE_URL`:

```ts
export const GAMELIST_BASE_URL = "https://m.np.playstation.com/api/gamelist";
```

The function importing this can tack on the other stuff, like:

```ts
const url = buildRequestUrl(GAMELIST_BASE_URL, "/v2/users", ...);
```

File: src/user/getUserPlayedTime.ts [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import { USER_TIME_PLAYED_URL } from "./USER_TIME_PLAYED_URL";
+
+type GetUserFriendsAccountIdsOptions = Pick<AllCallOptions, "limit" | "offset">;
+
+/**
+ * A call to this function will retrieve the list of playtime of various
+ * PS4 and PS5 titles. If the user cannot be found (either due to
+ * non-existence or privacy settings), an error will be thrown.
+ *
+ *  To find a user's `accountId`, the `makeUniversalSearch()` function can be used.
```
Comment (wescopeland) [inline src/user/getUserPlayedTime.ts:17]:
```suggestion
 * To find a user's `accountId`, the `makeUniversalSearch()` function can be used.
```

File: src/user/getUserPlayedTime.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @param authorization An object containing your access token, typically retrieved with `exchangeCodeForAccessToken()`.
+ * @param accountId The account whose trophy list is being accessed. Use `"me"` for the authenticating account.
```
Comment (wescopeland) [inline src/user/getUserPlayedTime.ts:20]: Options should be documented as well.

File: src/user/getUserPlayedTime.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  accountId: string,
+  options?: Partial<GetUserFriendsAccountIdsOptions>
+): Promise<GetUserPlayedTimeResponse> => {
+  const url = buildRequestUrl(
+    USER_TIME_PLAYED_URL,
+    "/:accountId/titles",
+    options,
+    {
+      accountId
+    }
```
Comment (wescopeland) [inline src/user/getUserPlayedTime.ts:33]:
```suggestion
    { accountId }
```

nit; Prettier should cooperate here

**Review by wescopeland — COMMENTED** [review]
Nice first foray here, left a bit of feedback. We also need:

- Tests (`yarn test`)
- Docs site updates (in the _website_ directory)
- _README.md_ update

File: src/models/game-media.model.ts (hunk 2 above)
Comment (TheYuriG) [reply src/models/game-media.model.ts:3]: I don't know how, I haven't seen a single game with data in those arrays yet. I could strongly type them if I found any examples, maybe a more recent game or a game published by a big company (like Ubisoft or Sony)?

File: src/models/game-media.model.ts (hunk 3 above)
Comment (TheYuriG) [reply src/models/game-media.model.ts:10]: You are correct, it's an array of objects with those properties. I'll make that change soon.

File: src/models/user-played-time-response.model.ts (hunk 6 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:14]: Sure, I'll add some and send for review again.

File: src/models/user-played-time-response.model.ts (hunk 9 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:23]: I never wrote typescript before this PR so I can't really say. I assume this is the strong typing you prefer to have?

File: src/models/user-played-time-response.model.ts (hunk 10 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:26]: It would probably cause issues (unless we mass-test to fetch several examples and narrow down the edge cases), the format for this is very erratic. I would assume this is not standardized internally somehow?

File: src/models/user-played-time-response.model.ts (hunk 13 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:38]: They should be ISO Strings because in my implementation I convert them to actual `Date` objects and then get their time in seconds with `.getTime()`.

For reference:

```
//? Instantiate time played data
const firstPlayed = Math.floor(
    new Date(title.firstPlayedDateTime ?? 0).getTime() / 1000
);
const lastPlayed = Math.floor(
    new Date(title.lastPlayedDateTime ?? 0).getTime() / 1000
);
```

File: src/models/user-played-time-response.model.ts (hunk 14 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:43]: My bad, forgot to do that when I was drafting the PR. All playDuration strings follow this format:

`PT` + `{Number of hours, if any}H` + `{Number of minutes, if any}M` + `{Number of seconds, if any}S`

How do you suggest improving this comment?

File: src/user/USER_TIME_PLAYED_URL.ts (hunk 15 above)
Comment (TheYuriG) [reply src/user/USER_TIME_PLAYED_URL.ts:1]: Yeah, that can be done on my next pass.

File: src/models/game-media.model.ts (hunk 2 above)
Comment (wescopeland) [reply src/models/game-media.model.ts:3]: If that's the case, let's use `unknown[]`.

File: src/models/user-played-time-response.model.ts (hunk 9 above)
Comment (wescopeland) [reply src/models/user-played-time-response.model.ts:24]: Yes, if the three possible values are known we should strongly-type them.

File: src/models/user-played-time-response.model.ts (hunk 13 above)
Comment (wescopeland) [reply src/models/user-played-time-response.model.ts:39]: Sorry, I am unclear, can you explain this another way?

Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:39]: Guess I'm overcomplicating it. It's an ISO string like the last played data that you get when fetching the user games. Same data type.

File: src/models/user-played-time-response.model.ts (hunk 14 above)
Comment (TheYuriG) [reply src/models/user-played-time-response.model.ts:44]: Don't know if I wasn't clear or if I just don't understand what strong typing is, so I'm just going to paste a few of the playtime durations you can get from this:
```
playDuration: 'PT1H14M10S'
playDuration: 'PT10H28M12S'
playDuration: 'PT1H52M23S'
playDuration: 'PT2M38S'
playDuration: 'PT7M30S'
playDuration: 'PT69H17M26S'
```

All of these are from the user `pandaswithbombs`, if you want to double check it.
Does this data fits the format that you suggested? If so, I'll just commit what you said.

Comment (wescopeland) [reply src/models/user-played-time-response.model.ts:44]: I have done some research on this value and it seems it's actually a technical standard: an ISO 8601 duration string. To be honest, this is the first time I've ever come across this.

After reading a bit more about the spec, we probably can't implement a strong type for this and will have to use `string`. However, we should document on the field explicitly that it's an ISO 8601 duration string.

ref: https://tc39.es/proposal-temporal/docs/duration.html

**Comment by wescopeland** [comment]
Hi! This PR has exceeded 30 days of inactivity. I will be closing it as a courtesy. Please feel free to re-open it at any time you would like to revisit this work.
